src/cluster_linearize.h

```diff
@@ -461,96 +461,123 @@ std::vector<FeeFrac> ChunkLinearization(const DepGraph<SetType>& depgraph, std::
- *     topological.
- * - Splitting:
- *   - Whenever an active dependency d exists, making it inactive will result in the chunk it is in
- *     splitting in two: a bottom chunk (which d is from) and a top chunk (which d is to). This is
- *     the case because no other active dependency between the top and bottom can exist; if it did,
- *     it would form a cycle together with d.
- *   - An active dependency can be deactivated whenever its would-be top chunk has strictly higher
- *     feerate than its would-be bottom chunk.
- *   - Splitting generally helps making a forest's output linearization better, but can result in
- *     it becoming non-topological, necessitating merging steps.
+ * We define three quality properties the state can have, each being stronger than the previous:
+ *
+ * - acyclic: The state is acyclic whenever no cycle of active dependencies exists within the
+ *            graph, ignoring the parent/child direction. This is equivalent to saying that within
+ *            each chunk the set of active dependencies form a tree, and thus the overall set of
+ *            active dependencies in the graph form a spanning forest, giving the algorithm its
+ *            name. Being acylic is also equivalent to every chunk of N transactions having exactly
+ *            N-1 active dependencies.
+ *
+ *            For example in a diamond graph, D->{B,C}->A, the 4 dependencies cannot be
+ *            simultaneously active. If at least one is inactive, the state is acyclic.
+ *
+ *            The algorithm maintains an acyclic state at *all* times as an invariant. This implies
+ *            that activating a dependency always corresponds to merging two chunks, and that
+ *            deactivating one always corresponds to splitting two chunks.
+ *
+ * - topological: We say the state is topological whenever no inactive dependency exists between
+ *                two distinct chunks such that the child chunk has higher or equal feerate than
+ *                the parent chunk.
+ *
+ *                The relevance is that whenever the state is topological, the produced output
+ *                linearization will be topological too (i.e., not have children before parents).
+ *                Note that the "or equal" part of the definition matters: if not, one can end up
+ *                in a situation with mutually-dependent equal-feerate chunks that cannot be
+ *                linearized. For example C->{A,B} and D->{A,B}, with C->A and D->B active. The AC
+ *                chunk depends on DB through C->B, and the BD chunk depends on AC though D->A.
+ *                Merging them into a single ABCD chunk fixes this.
+ *
+ *                The algorithm attempts to keep the state topological as much as possible, so it
+ *                can be interrupted to produce an output whenever, but will sometimes need to
+ *                temporarily deviate from it when improving the state.
```

Comment:
> temporarily deviate from it when improving the state.

I'm expecting each `Improve()` call to result in a topological state, assuming one exists, and that this means during a call